### PR TITLE
[feat]: [PLG-2661]: iterator to provision free monthly credits for each CI module license

### PR DESCRIPTION
**Pull request by kapil-harness**
## Describe your changes
iterator to provision free monthly credits for each CI module license

Scenarios tested:
1. No credits should be provisioned if its not 1st day of the month
2. No credits should be provisioned if credits already provisioned for that month
3. Provision 2k credits if 1st day of the month & no credits already provisioned for that month

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52060)

### CODE REVIEW

**Comment by kapil-harness** [comment]
trigger uts

**Comment by kapil-harness** [comment]
trigger uts

File: 930-ng-license-manager/src/main/java/io/harness/credit/schedular/ProvisionMonthlyCICreditsHandler.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  private boolean isCreditAlreadyProvisionedForCurrentMonth(CIModuleLicense ciModuleLicense) {
+    List<CreditDTO> creditDTOS = creditService.getLatestFreeAndActiveCredits(ciModuleLicense.getAccountIdentifier());
+    if (creditDTOS.isEmpty()) {
+      return false;
+    }
+
+    return Calendar.getInstance().get(Calendar.MONTH) == getMonthOfPurchaseOfLatestCredit(creditDTOS);
+  }
+
+  private int getMonthOfPurchaseOfLatestCredit(List<CreditDTO> creditDTOS) {
```
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/credit/schedular/ProvisionMonthlyCICreditsHandler.java:117]: let's just name this as `getLatestPurchaseTime` ?

**Review by neelam-harness — APPROVED** [review]
minore refactoring comment, lgtm overall!

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52060)* status: 9 file(s) to review. 1 change requests. 1 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/credit/services/impl/CreditServiceImpl.java` line 64 at r2](https://reviewable.io/reviews/harness/harness-core/52060#-NfFef_7CUQ76nEx1EuB:-NfFef_7CUQ76nEx1EuC:b1sj709) ([raw file](https://github.com/harness/harness-core/blob/a34b2701a210fa9eb1bea1043273cb499dd1f4c0/930-ng-license-manager/src/main/java/io/harness/credit/services/impl/CreditServiceImpl.java#L64)):*
> ```Java
> 
>   @Override
>   public List<CreditDTO> getLatestFreeAndActiveCredits(String accountIdentifier) {
> ```

It would be good to make this service method configurable in terms of credit status. It should just ask for list of credit status applicable and return results for the same.

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52060)* status: 9 file(s) to review. 1 change requests. 1 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/repositories/CreditRepository.java` line 28 at r2](https://reviewable.io/reviews/harness/harness-core/52060#-NfFeqN2DSrBfuVwHy-7:-NfFeqN2DSrBfuVwHy-8:b3ng7rw) ([raw file](https://github.com/harness/harness-core/blob/a34b2701a210fa9eb1bea1043273cb499dd1f4c0/930-ng-license-manager/src/main/java/io/harness/repositories/CreditRepository.java#L28)):*
> ```Java
>   List<Credit> findByAccountIdentifier(String accountIdentifier);
> 
>   List<Credit> findByAccountIdentifierAndCreditTypeAndCreditStatus(
> ```

ensure its indexed

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52060)* status: 9 file(s) to review. 1 change requests. 1 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/credit/schedular/ProvisionMonthlyCICreditsHandler.java` line 143 at r2](https://reviewable.io/reviews/harness/harness-core/52060#-NfFfK9zC6qvTpqzZE_s:-NfFfK9zC6qvTpqzZE_t:b-tgmmjg) ([raw file](https://github.com/harness/harness-core/blob/a34b2701a210fa9eb1bea1043273cb499dd1f4c0/930-ng-license-manager/src/main/java/io/harness/credit/schedular/ProvisionMonthlyCICreditsHandler.java#L143)):*
> ```Java
>   private static Calendar getStartOfNextMonth() {
>     Calendar calendar = Calendar.getInstance();
>     calendar.setTimeZone(TimeZone.getTimeZone(GMT_TIMEZONE));
> ```

I think we should not set timezone here as we aren't setting it anywhere

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 2 of 5 files at r1, 2 of 6 files at r2, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52060)* status: 10 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by neelam-harness — APPROVED** [review]
Let's get a sign-off from DBRE as well wrt indexes for this change. 
Otherwise, lgtm!

**Comment by kapil-harness** [comment]
trigger review

**Pull request merged**